src/addons/vue-directives.ts

```diff
@@ -0,0 +1,114 @@
+  const directivesArray = Array.isArray(directives) ? directives : [directives]
+  const directivesPromise = Promise
+    .all(directivesArray.map(async (preset) => {
+      return [preset, await preset.directives] as const
+    }))
+    .then((entries) => {
+      const map = new Map<string, DirectiveType>()
+      for (const [preset, directive] of entries) {
+        if (Array.isArray(directive)) {
+          for (const entry of directive) {
+            map.set(entry.directive, [entry, preset])
+          }
+        }
+        else {
+          map.set(directive.directive, [directive, preset])
+        }
+      }
+      return map
+    })
+
+  const self = {
+    async transform(s, id) {
+      if (!s.original.includes('_ctx.') || !s.original.match(contextRE))
+        return s
+
+      const directivesMap = await directivesPromise
+      let targets: Import[] = []
+      // We have something like this:
+      // var/const _directive_click_outside = _resolveDirective("click-outside");?
+      // We need to remove the declaration and replace it with the corresponding import
+      // extracting the symbol and the directive name from the declaration
+      const matches = Array
+        .from(s.original.matchAll(directiveRE))
+        .sort((a, b) => b.index - a.index)
+        .reduce((acc, regex) => {
+          const [all, symbol, name] = regex
+          const directiveName = `v-${name}`
+          const entry = directivesMap.get(directiveName)
+          if (!entry)
+            return acc
+
+          const [directive, preset] = entry
+          const importName = directive.name
+          const asStmt = directive.as
+          // remove the directive declaration
+          s.overwrite(
+            regex.index,
+            regex.index + all.length,
+            '',
+          )
+          if (importName !== 'default') {
+            // add named import
+            targets.push({
+              ...preset,
+              name: asStmt ?? importName,
+              as: symbol,
+            })
+          }
+          else {
+            // add default export
+            targets.push({
+              ...preset,
+              name: 'default',
+              as: symbol,
+            })
+          }
+          return acc + 1
+        }, 0)
+
+      if (!matches)
+        return s
+
+      // remove resolveDirective import
+      s.replace(contextText, '')
+
+      for (const addon of this.addons) {
+        if (addon === self)
+          continue
+
+        targets = await addon.injectImportsResolved?.call(this, targets, s, id) ?? targets
+      }
+
+      let injection = stringifyImports(targets)
+      for (const addon of this.addons) {
+        if (addon === self)
+          continue
+
+        injection = await addon.injectImportsStringified?.call(this, injection, targets, s, id) ?? injection
+      }
+
+      s.prepend(injection)
+
+      return s
+    },
```

Comment:
I tried to see if I could add the directive types, it seems the following code works partially, the types are picked up but there's no autocompletion for me.

```suggestion
    },
    async declaration(dts, options) {
      const dirs = Array.from(await directivesPromise)

      const items = dirs.map(([_, dir]) => {
        const from = options?.resolvePath?.({ ...dir[0], from: dir[1].from }) || ''
        // `camelCase` function from 'scule'
        // NOTE: import path should be absolute, adding `.` temporarily since it's relative
        return `${camelCase(dir[0].directive)}: typeof import('.${from}')['${dir[0].name}']`
      })
        .filter(Boolean)
        .sort()

      const extendItems = items.map(i => `    ${i}`).join('\n')
      return `${dts}

declare module 'vue' {
  interface GlobalDirectives {
${extendItems}
  }
}`
    },
```

Reply:
Directives are not components, should be in global.

Your suggestion will not work if we have abother addon augmenting vue module

Reply:
In fact, we should review the existing vue addon...

Reply:
so, we only need to add them in the `ComponentCustomProperties`, resolving this

ts allows us to include multiple `vue` module augmentation even using the same `ComponentCustomProperties`  interface in the same dts file

Reply:
```suggestion
    },
    async declaration(dts, options) {
      const dirs = Array.from(await directivesPromise)

      const items = dirs.map(([_, dir]) => {
        const from = options?.resolvePath?.({ ...dir[0], from: dir[1].from }) || ''
        // `camelCase` function from 'scule'
        // NOTE: import path should be absolute, adding `.` temporarily since it's relative
        return `${camelCase(dir[0].directive)}: typeof import('.${from}')['${dir[0].name}']`
      })
        .filter(Boolean)
        .sort()

      const extendItems = items.map(i => `    ${i}`).join('\n')
      return `${dts}

declare module 'vue' {
  interface ComponentCustomProperties {
${extendItems}
  }
}
`
```